### DISCUSSION TITLE
Why GetInput() works so bad for Fork, Foreach

### DISCUSSION THREAD

**Discussion by kimdimasan** [elsa-workflows/elsa-core#4570] [Q&A]
Is it only me or is it really strange that I have to write custom code to save the first run ctx.GetInput() value into the Variables for Fork/ForEach because otherwise it returns different value every time (previously running stage). It sounds counter-intuitive. 

The documentation is really bad for this project, and the code is lacking comments. Also the samples are almost always for some simplified cases where the input is static or something like that. It's really hard to work with it.

**Comment by sfmskywalker** [comment]
Hi @kimdimasan,

Thank you for taking the time to share your feedback and experiences with Elsa. We understand your concerns and appreciate your patience as we continuously work to improve the library.

Regarding the `GetInput()` method, it's important to note that it represents the last output provided by the last activity that executed. This can indeed lead to some unexpected behaviors when used within loops, such as Fork and Foreach. We acknowledge that this might not be immediately clear, and it's something we aim to address in our documentation and examples.

Speaking of documentation, we recognize that there are areas in need of improvement, especially in terms of clarity and comprehensiveness. We are actively working on enhancing the documentation for Elsa 3, and we are also making strides in adding more XML comments to the codebase to provide better inline guidance. Your feedback is invaluable in this process, as it helps us identify the areas that need the most attention.

If you could provide more details about your specific workflow and how you are using `GetInput()`, we would be more than happy to offer more tailored advice and guidance.

Once again, thank you for your feedback. It plays a crucial role in helping us make Elsa a better tool for everyone.

**Comment by kimdimasan** [comment]
[sfmskywalker](https://github.com/sfmskywalker)
For example previous activity is giving me a Collection of elements.
The next activity is ForEach, so I set the items like `.ForEach(ctx => ctx.GetInput<ICollection<Type>>(), iterate => iterate...` The problem now is that the first part with GetInput is called every iteration instead of just once and GetInput is always different (the result of the last activity).  What is the best approach to make sure that ForEach.Items are always the same instead of changing every iteration? For now I am using SetState/GetState to save the first GetInput value (for the first interation) and use it for next iterations, but not sure if it will work for nested loops.

**Reply by sfmskywalker** [reply]
The safest way to handle this is to access activity output directly and or to save that into a workflow variable. which you then access. Elsa 2 has some limitations in this regard, which is fixed with Elsa 3. In fact, these limitations in architecture were a big driving factor to rewrite everything from the ground up. If you can migrate to V3, I highly recommend it.

**Comment by shanwayne** [comment]
@sfmskywalker One other expectation at the moment, is that input in a Flowchart activity would be available for all subsequent activities but I'm finding that Input is empty for subsequent activities. If I understand this the intention is for Flowchart input to flow into the starting activity input and if the input is needed then it needs to be manually propagated. Is that correct?

Note: I'm currently experimenting with using with using Elsa v3 in the external app integration scenario with custom activities.

**Reply by sfmskywalker** [reply]
Can you go into more detail about how you are sending input to a flowchart activity? In any case, workflow input should be accessible to all activities in that workflow, at least until the first suspension point.

**Reply by shanwayne** [reply]
@sfmskywalker We have a custom activity that interacts with our internal API then suspends itself and waits to be resumed. The expectation is after resuming the workflow input should still be available for subsequent activities.

**Reply by sfmskywalker** [reply]
I see. The difficulty here is that when resuming a workflow instance, new input may be available. This means that as of now, Elsa essentially only provides the input for as long as the workflow is executing. When the workflow is resumed, it receives new input (if any) and old input is discarded.

That being said, I do think your expectation makes sense, and perhaps we could design it as follows:

- Input is always maintained by the workflow instance
- When a workflow is resumed and receives new input, it is added to the list of inputs.
- If new input is received that uses the same key as an existing input, it is simply overwritten.
- To handle non-serialisable input values like services or other non-serializable objects like Streams, it is up to the workflow creator to make sure these inputs use the appropriate storage provider.

I'll create an issue for this change request 👍🏻

**Reply by sfmskywalker** [reply]
https://github.com/elsa-workflows/elsa-core/issues/4705

**Reply by shanwayne** [reply]
Yes, I agree with this plan. This would be an improvement, thank you!
